enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,616 @@
+ran MicroShift is present.
+
+To allow for such operations, we need to define how we'll
+achieve that goal. We can define several areas we need to focus on:
+backing up and restoring MicroShift's data, handling version changes
+and its consequences such as migrating underlying data between schema
+versions, defining a mechanism for allowing or blocking upgrades between
+certain version of MicroShift.
+
+### User Stories
+
+* As a MicroShift administrator, I want to safely update MicroShift
+  so that I can get bug fixes, new features, and security patches.
+* As a MicroShift administrator, I want automated backups of 
+  MicroShift data so in case of a greenboot failure system will rollback
+  and backup will be restored.
+
+### Goals
+
+Goal of the enhancement is to describe implementation roadmap for
+integrating MicroShift with ostree and greenboot in order to provide
+functionality to:
+- Safely update MicroShift version (by backing up the data and
+  restoring it in case of rollback)
+- Migrating internal data (like Kubernetes storage or etcd schema) to
+  newer version
+- Block upgrades in case of version bump being too big
+
+Design aims to implement following principles:
+- Keep it simple, optimize later
+- MicroShift does not own the OS or host
+- MicroShift and all its components are version, upgraded, and rolled back together
+- Be defensive, fail fast
+- Rely on outside intervention as a last resort
+
+### Non-Goals
+
+* Building allowed/blocked version migration graph
+* Handling readiness, and backup and rollback of 3rd party applications
+  (although end user documentation should be provided)
+* Defining updateability for non-ostree systems is left to a future enhancement
+* Protecting against data corruption
+
+## Proposal
+
+### Workflow Description
+
+#### User perspective
+
+**MicroShift administrator** is a human responsible for preparing
+ostree commits and scheduling devices to use these commits.
+
+1. MicroShift administrator prepares a new ostree commit
+1. MicroShift administrator schedules device to reboot and use new ostree commit
+1. Device boots to new commit
+1. Operating System, greenboot, and MicroShift take actions without any additional intervention
+
+#### Implementation perspective
+
+**First boot into commit with MicroShift**
+
+1. First ostree commit is installed
+1. First ostree commit boots
+1. MicroShift starts
+   - No backup action (did not previously run, so there is nothing to backup, and did not previously fail so no restore needed)
+1. MicroShift startup succeeds
+1. All greenboot checks pass
+1. Greenboot runs green scripts
+   - MicroShift green script sets backup mode to "backup"
+
+**Simple host reboot (cont of previous flow)**
+
+1. First ostree commit shuts down
+1. First ostree commit boots
+1. MicroShift starts up
+   - Backup mode is "backup",
+   - Backup script runs, creating a backup compatible with the first ostree commit
```

Comment:
Should we state that we overwrite the existing backup on every backup operation?

Reply:
Note that we may not want to create / overwrite backup on greenboot reboot attempts. 
So, a backup should be taken on the first attempt when it does not exist. Otherwise, there is a chance of backup "corruption" in-between these attempts.
How do we do that?

Reply:
Greenboot reboots will happen due to unhealthy boot and for such red scripts will run. Out red script will set mode "restore", so on first boot we do a backup, if it fails and reboots, we'll restore that backup and basically retry (e.g. migrate, start again, etc.)

Reply:
Right. Do we want to spell it out that we're starting from a "clean slate" on every failed boot?

Reply:
I added a paragraph in `### Backup and restore of MicroShift data`. PTAL